### DISCUSSION TITLE
How robust is the cp command?

### DISCUSSION THREAD

**Discussion by MerryMurray** [nushell/nushell#7427] [Q&A]
### Question

The coreutils cp command does not handle large folders well.
The rust app lms handles huge folder copies well. I just transferred
2.6 terrabytes from one drive to another in 20 hours with the
single command: lms cp /folder/ondrivesource /folder/ondrivetarget.
Hoping cp command does or will be this robust.

Murray

**Comment by fdncred** [comment]
never heard of lms, do you have a url for it?

**Comment by rgwood** [comment]
`cp` should be robust in the sense that it copies files reliably. Let us know if you have specific feature requests or bug reports for `cp`.

**Comment by theAkito** [comment]
> transferred 2.6 terrabytes from one drive to another in 20 hours

Isn't this a classic `rsync` job? I generally wouldn't recommend anyone to use any `cp` for anything, that has a complex permission setup or a huge amount of data.

**Comment by MerryMurray** [comment]
I wouldn't use standard cp either. Rsync? The target drive is empty.Lms is designed for this. It did the 2.6g transfers  flawlessly twiceas well as many other multi gigabyte transfers.
That is why I suggested it.  Because it is way better than standardcopy. 

    On Tuesday, December 20, 2022 at 12:07:30 a.m. PST, theAkito ***@***.***> wrote:  
 
 



transferred 2.6 terrabytes from one drive to another in 20 hours


Isn't this a classic rsync job? I generally wouldn't recommend anyone to use any cp for anything, that has a complex permission setup or a huge amount of data.

**Reply by theAkito** [reply]
> Rsync? The target drive is empty.Lms is designed for this. It did the 2.6g transfers  flawlessly twiceas well as many other multi gigabyte transfers.

You specifically said the following.

> The coreutils cp command does not handle large folders well.

Which is, why I recommended `rsync`, because once `cp` is becoming flakey, you would change to `rsync`. For example, with such huge folders it is very common, that the copy operation is interrupted for whatever reason.
With `rsync` you can just restart the process, considering you already have the right options set. With `cp` it isn't always expected to offer all options for such situations, because it's just meant to copy relatively simple things.
Since you said it "does not handle large folders well", I assume your folder wasn't simple enough.

**Comment by MerryMurray** [comment]
Hi
For the nushell cp command what is the maximum folder sizeyou would use ?
    On Tuesday, December 20, 2022 at 09:28:57 p.m. PST, theAkito ***@***.***> wrote:  
 
 



Rsync? The target drive is empty.Lms is designed for this. It did the 2.6g transfers flawlessly twiceas well as many other multi gigabyte transfers.


You specifically said the following.


The coreutils cp command does not handle large folders well.


Which is, why I recommended rsync, because once cp is becoming flakey, you would change to rsync. For example, with such huge folders it is very common, that the copy operation is interrupted for whatever reason.
With rsync you can just restart the process, considering you already have the right options set. With cp it isn't always expected to offer all options for such situations, because it's just meant to copy relatively simple things.
Since you said it "does not handle large folders well", I assume your folder wasn't simple enough.

**Reply by theAkito** [reply]
> Hi For the nushell cp command what is the maximum folder sizeyou would use?

I did not mention any specific size, because it depends on the situation. Sometimes you stop at 5GB, other times at 100MB, but I guess a rule of thumb is, that if you cross the TB line on an average desktop PC for a single end-user, then you are out of `cp` territory.
It is a rule of thumb & all depends on the situation, but this rule of thumb most likely applies to all single end-user PC systems.

It becomes more clear, when you try to `cp` something & it doesn't work out. If it doesn't work out, you should switch to `rsync` with proper options, according to the situation.
